doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+   An error may not be externally observable by the user of the system.
+3. A **failure** occurs if the system observably acts in contradiction to its specification.
+
+This reasonably assumes that errors are always caused by faults. If these errors are not detected
+they can cause failures. By detecting and handling errors failures are avoided.
+
+The goal of this document is to define error handling requirements.
+
+## Error Detection vs. Error Handling
+
+To handle an error it obviously needs to be detected first. To do so, the code must contain conditionals that check
+for inconsistent states, unavailability of resources, return codes of system calls etc.
+This means that the error has to be anticipated by the system designer,
+i.e. we cannot handle unforseen errors since they are not detected by definition.
+
+Error handling is the execution of handling code once the error was detected.
+This ranges from simple logging to sophisticated recovery or shutdown code and depends on the error in question.
+We distinguish between recoverable and non-recoverable error handling.
+
+### Recoverable Error Handling
+
+A recoverable error handling strategy allows the system to continue operation after the handling code
+was executed successfully. Exceptions are an example of such a mechanism but also any handler code
+that does not result in termination. If the handling code itself fails (e.g. insufficient resources)
+no recovery is possible and the system terminates.
+
+### Non-recoverable Error Handling
+
+A non-recoverable error handling strategy will always call terminate after the handling code was executed.
+Handler code may be used to initiate a graceful shutdown. An example are (strict) asserts.
+
+## Requirements
+
+In the following we list the requirements of the Error Handling (EH) strategy. The term platform
+refers to the usage environment of the error handler. There can be various platforms,
+e.g. for regular operations, testing, different OS etc.
+
+### Functional
+
+1. The EH shall support error codes.
+    - configurable on a platform basis
+    - errors from different platforms shall not be usable (checked at compile time)
+    - uniqueness is not enforced (can be done by the layer above if desired)
+    - error codes shall be optional (if none is used, it is mapped to some default error code)
+1. The EH shall support multiple error levels.
+    - the individual error levels are defined by the platform
+    - there has to be at least a FATAL error level (most severe error level)
+    - the error levels shall be ordered from least severe to most severe
+    - if the code requires an ERROR level that is not supported by the platform it shall not compile
+1. The EH shall support a non-recoverable error handling strategy in case of FATAL errors.
+    - non-recoverable errors shall call terminate after optional handling and logging code has been executed
+1. The EH shall support a recoverable error handling strategies for non-FATAL errors.
```

Comment:
This can become very compex, therefore I propose to only support the `cxx::expected` approach for non recoverable errors.

Reply:
No, I think this would be a mistake and too restrictive. It also allows the possibility to easily upgrade into a "WARNING" is handled like a FATAL error mode without changing the code in all places.

Or downgrade to "WARNING" means logging only.

The problem with `expected` is that without propagation it is useless. The other problem is that personally I find it cumbersome to use and so it should only be used it where propagation is required.

`expected` and the error detection mechanism do not serve the same purpose.

Reply:
I agree with @elfenpiff that for recoverable errors we should use `cxx::expected`. This is the most idiomatic way. However, those still need to be reported and stored in e.g. in a persitency module. What do you think about my proposal few threads above to call the `errorReporter/errorHandler` from `cxx::error`? I don't like our triplet of

1. log entry
2. `errorHandler` call
3. return of `cxx::error`

That's quite noisy and inconvenient.